clients/src/main/java/org/apache/kafka/common/utils/Time.java

```diff
@@ -53,4 +57,24 @@
      */
     void sleep(long ms);
 
+    /**
+     * Schedule a callable to be executed in the future.
+     *
+     * This is useful for testing delayed work that we would like to place on an
+     * ScheduledExecutorService.  If we just invoked ScheduledExecutorService#schedule
+     * directly, the executor would always use the system clock.  This method allows
+     * us to substitute our own clock where appropriate.
+     *
+     * Note that the Callable may not be queued on the executor until the designated
+     * time arrives.  So this method is not a perfect substitute for the
+     * ScheduledExecutorService#schedule() method.
+     *
+     * @param executor      The executor to use.
+     * @param callable      The callable to execute.
+     * @param delayMs       The delay to use, in milliseconds.
+     * @param <T>           The return type of the callable.
+     * @return              A future which will complete when the callable is finished.
+     */
+    public <T> Future<T> schedule(final ScheduledExecutorService executor,
+                                  final Callable<T> callable, long delayMs);
```

Comment:
I think the difference here is that we are reusing a standard Java interface, ScheduledExecutorService, rather than creating our own class, like KafkaScheduler.  I think the standard Java interface is more powerful and more familiar to most programmers.  But because it's a bigger interface, it's pretty hard to create a MockScheduledExecutorService.

In general, I'm not sure how to modify this to separate Scheduler from Time.  It seems like it might get pretty messy... we'd have to add Scheduler objects to the constructors of things.  But the Scheduler objects would have to take ownership of a thread created inside the thing, not passed from outside.

Reply:
An alternative may be to extend `Time` with a new interface in `tools` which does scheduling as well? Its usage would be the same as the current code, but the interface name could indicate that it is used for scheduling as well. And it could reuse the `Time` part of the implementation from `SystemTime` and `MockTime`.

Reply:
I'm looking into creating a MockScheduledExecutorService.  If I can do it, it will be more generally useful than the other options.  I have it almost written